src/pdf/filter.rs

```diff
@@ -0,0 +1,111 @@
+use mupdf_sys::*;
+
+use std::ffi::CStr;
+use std::os::raw::{c_char, c_void};
+use std::ptr;
+
+use crate::pdf::PdfDocument;
+use crate::{Image, Matrix, Rect};
+
+#[derive(Clone, Copy)]
+pub struct PdfFilterOptions {
+    pub(crate) inner: *mut pdf_filter_options,
+}
+
+// Callback types
+// TODO: I don't think this is necessary for anything.
+pub type ImageFilter = fn(ctm: &Matrix, name: &str, image: &Image) -> Image;
+pub type TextFilter = fn(ucsbuf: i32, ucslen: i32, trm: &Matrix, ctm: &Matrix, bbox: &Rect) -> i32;
+pub type AfterTextObject = fn(doc: &PdfDocument, chain: &pdf_processor, ctm: &Matrix);
+pub type EndPage = fn();
+
+impl PdfFilterOptions {
+    pub(crate) unsafe fn from_raw(ptr: *mut pdf_filter_options) -> Self {
+        Self { inner: ptr }
+    }
+
+    pub fn ascii(&self) -> bool {
+        unsafe { (*self.inner).ascii != 0 }
+    }
+
+    pub fn set_ascii(&mut self, value: bool) -> &mut Self {
+        unsafe {
+            (*self.inner).ascii = if value { 1 } else { 0 };
+        }
+        self
+    }
+
+    pub fn recurse(&self) -> bool {
+        unsafe { (*self.inner).recurse != 0 }
+    }
+
+    pub fn set_recurse(&mut self, value: bool) -> &mut Self {
+        unsafe {
+            (*self.inner).recurse = if value { 1 } else { 0 };
+        }
+        self
+    }
+
+    pub fn sanitize(&self) -> bool {
+        unsafe { (*self.inner).sanitize != 0 }
+    }
+
+    pub fn set_sanitize(&mut self, value: bool) -> &mut Self {
+        unsafe {
+            (*self.inner).sanitize = if value { 1 } else { 0 };
+        }
+        self
+    }
+
+    pub fn instance_forms(&self) -> bool {
+        unsafe { (*self.inner).instance_forms != 0 }
+    }
+
+    pub fn set_instance_forms(&mut self, value: bool) -> &mut Self {
+        unsafe {
+            (*self.inner).instance_forms = if value { 1 } else { 0 };
+        }
+        self
+    }
+
+    pub fn set_image_filter(
+        &mut self,
+        wrapper: fn(&Matrix, &str, &Image) -> Option<Image>,
+    ) -> &mut Self {
+        use once_cell::sync::OnceCell;
+
+        static WRAPPER: OnceCell<fn(&Matrix, &str, &Image) -> Option<Image>> = OnceCell::new();
+        // TODO: for some reason I can't use unwrap here??
+        let _ = WRAPPER.set(wrapper);
+
+        unsafe extern "C" fn image_filter_callback(
+            _ctx: *mut fz_context,
+            _opaque: *mut c_void,
```

Comment:
I think the `opaque` argument should be passed into the callback? Otherwise it means we miss some context information in the callback.

In upstream it is used in some filters, for example https://github.com/ArtifexSoftware/mupdf/blob/a230d9c4255d4fefc6b3e71ba41447f5b0620e82/source/pdf/pdf-clean.c#L547

Reply:
Yup, I missed that.

Reply:
Where exactly should I pass that? I don't really understand why it's assigning its value to a pdf_page. And with the mupdf-rs bindings it's not like you can pass a pointer to void anywhere as a user.

Reply:
Looks like `opaque` is only set once here: https://github.com/ArtifexSoftware/mupdf/blob/a230d9c4255d4fefc6b3e71ba41447f5b0620e82/source/pdf/pdf-clean.c#L756

Reply:
I'm starting to think that `opaque` is the `void*` we can use to pass our own parameters... But that would be bad naming... I've asked in their Discord and I'll update this once I know more.

Reply:
Yup, `opaque` is meant to be used to pass the data around. Now that I've fixed that I've got one last doubt: perhaps we should be using the `_ctx` value instead of ignoring it? Not sure how that works.